### ISSUE TITLE
Opening file on tramp remote ftp after initialization inserts error in file buffers

### ISSUE THREAD

**Issue by cjohansson** [magit/magit#3684]
**(1) what behavior you expected**

Opening a file should not alter it's buffer by default

**(2) what behavior you observed**

The contents "Unknown remote command: git" is inserted

**(3) and how we can reproduce the issue.**

Start emacs and load magit, since I lazy-load it when running magit-status I run magit-status. Now open a file over tramp ftp and at the top of the file you should see the error. I only get this error when magit is loaded.

**Also post the output of M-x magit-version RET**

Magit 20181204.2004, Git te command: git, Emacs 27.0.50, darwin

I was unable to load all dependencies from emacs -Q using this:

```
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/dash-20180910.1856/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/git-commit-20181124.1111/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/magit-popup-20181204.2031/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/with-editor-20181113.1845/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/ghub-20181130.1001/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/magit-20181204.2004/"))
(require 'magit)
```

I get this error `Turning on magit-auto-revert-mode...done
byte-code: Cannot open load file: No such file or directory, graphql`

**Comment by tarsius** [comment]
> I get this error `Turning on magit-auto-revert-mode...done
byte-code: Cannot open load file: No such file or directory, graphql`

Start by fixing that: Magit currently needs `graphql`.

**Comment by cjohansson** [comment]
Alright got it working using this

```

(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/dash-20180910.1856/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/git-commit-20181124.1111/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/magit-popup-20181204.2031/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/with-editor-20181113.1845/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/ghub-20181130.1001/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/magit-20181204.2004/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/graphql-20180912.31/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/treepy-20180724.656/"))
(require 'magit)
```

Issue remains

**Comment by tarsius** [comment]
> Opening file on tramp remote ftp

You are using Tramps `ftp` method? That cannot work because Magit doesn't just need to edit files, it also needs to run the `git` executable. Try with the `ssh` method.

**Comment by cjohansson** [comment]
But it must be a bug that I can’t open files over FTP without buffers being modified once Magit has been loaded into Emacs? 

I understand that Magit can’t work on ftp hosts but it shouldn’t by default alter buffers with error messages?

**Comment by cjohansson** [comment]
I think this example better explains the issue

```elisp
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/dash-20180910.1856/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/git-commit-20181124.1111/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/magit-popup-20181204.2031/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/with-editor-20181113.1845/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/ghub-20181130.1001/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/magit-20181204.2004/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/graphql-20180912.31/"))
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/treepy-20180724.656/"))
(require 'magit)
(find-file "/ftp:anonymous@speedtest.tele2.net:/512KB.zip")
```

![skarmavbild 2018-12-06 kl 07 53 30](https://user-images.githubusercontent.com/924848/49566759-0e624a80-f92c-11e8-8565-4307d32d5549.png)

**Comment by tarsius** [comment]
I have no idea why this happens. Grepping Tramp for "Unknown remote command" (and possibly inserting debug statements or even `(error "gimme backtrace")` nearby) might give some insights.

**Comment by tarsius** [comment]
Just to make sure; I was trying to suggest that you should do that.

**Comment by cjohansson** [comment]
Alright, sure I can do that. 😃 The issue does not occur with the built-in GIT handler and I have temporarily been using that

**Comment by tarsius** [comment]
Looking at the screenshot again I see you are visiting `512KB.zip`. That's a binary file so I don't think it is strange that it looks strange in a text editor.

> Unknown remote command: git

So Magit+Tramp are not able to find `git`. That's a configuration issue, most likely.

Check the value of `magit-git-executable`. If it is an absolute patch, then check whether that is valid on the remote host. If it is "git" then you have to make sure that Tramp puts that on the path it searches.

**Comment by cjohansson** [comment]
Hi again, I checked with GNU Emacs maintainers and this issue is considered to belong to magit and not GNU Emacs. There should be a defense mechanism in magit that handles situations were `(process-file)` does not work properly i.e. when having a ftp connection you can't run remote processes. 

Is it possible to add some kind of conditional-logic checking if it's possible to run processes before calling `(process-file)`?

**Comment by cjohansson** [comment]
Same result for me, maybe the error depends on OS or Emacs-version? I use macOS 10.14.2 and GNU Emacs 27.0.50 (build 12, x86_64-apple-darwin18.2.0, NS appkit-1671.20 Version 10.14.2 (Build 18C54))
 of 2019-01-14

I can try to find a solution to this

![skarmavbild 2019-01-22 kl 07 47 04](https://user-images.githubusercontent.com/924848/51517090-f2523400-1e19-11e9-8cbe-c8c9ab5a63b0.png)

**Comment by tarsius** [comment]
> I can try to find a solution to this

Please try. Maybe turning on the debugger would help. Since it did not happen for me I don't know what feature exactly tried to use git. Once we know what exactly did it, a save-guard can be added.

**Comment by pietvo** [comment]
I ran into the same problem today. I administer a few websites through `ftp`, and when I opened a file with `/ftp:site:/filename:`, I got "Unknown remote command: git" twice inserted at the top of the file.
I traced it, and it gets there in this way:
The function `normal-mode` calls 
```
(run-hooks 'change-major-mode-after-body-hook
               'after-change-major-mode-hook))
```
`after-change-major-mode-hook` contains a.o. `magit-auto-revert-mode-enable-in-buffers`.
This calls `magit-turn-on-auto-revert-mode-if-desired`, which in turn calls `magit-toplevel`.
This tries to run the `git` command on the remote site, which fails, and then for some reason the error message is inserted in the bugger.

I solved the problem locally, by advising `magit-toplevel` to do nothing in an `tramp`-controlled buffer. It could be more subtle by allowing it for `ssh` or `sudo` accessed buffers and disallowing it for `ftp`, or some other scheme but this works for me for the time being. I think a more general solution would be to introduced a customisable predicate, maybe with a choice of presets.

Here is my code:
```
(defun no-tramp-magit-toplevel (orig-fun &rest args)
  (when (and buffer-file-name
		(not (tramp-tramp-file-p buffer-file-name)))
	(apply orig-fun args)))

(advice-add 'magit-toplevel :around #'no-tramp-magit-toplevel)

```
